src/rgw/rgw_data_sync.cc

```diff
@@ -4136,12 +4138,78 @@ static bool has_olh_epoch(RGWModifyOp op) {
   return op == CLS_RGW_OP_LINK_OLH || op == CLS_RGW_OP_UNLINK_INSTANCE;
 }
 
+class RGWBucketShardIsDoneCR: public RGWCoroutine {
+  RGWDataSyncCtx *sc;
+  RGWDataSyncEnv *sync_env;
+  rgw_bucket_sync_status bucket_status;
+  const rgw_raw_obj status_obj;
+  const uint64_t num_shards;
+  const int shard_id;
+  RGWObjVersionTracker objv_tracker;
+  std::optional<next_bilog_result>& next_log;
+
+public:
+  RGWBucketShardIsDoneCR(RGWDataSyncCtx *_sc, const std::string& _status_oid,
+                         int _num_shards, int _shard_id, std::optional<next_bilog_result>& _next_log)
+    : RGWCoroutine(_sc->cct), sc(_sc), sync_env(_sc->env),
+      status_obj(sync_env->svc->zone->get_zone_params().log_pool, _status_oid),
+      num_shards(_num_shards), shard_id(_shard_id), next_log(_next_log) {}
+
+  int operate() override
+  {
+    reenter(this) {
+      do {
+      // read bucket sync status
+      using ReadCR = RGWSimpleRadosReadCR<rgw_bucket_sync_status>;
+      yield call(new ReadCR(sync_env->async_rados, sync_env->svc->sysobj,
+                            status_obj, &bucket_status, false, &objv_tracker));
```

Comment:
each time we read/reread this status, we should do a couple checks before trying to apply our changes. this is where we have to handle the possible races with other shard completions, or with changes to the BucketSyncState from commands like 'bucket sync init' or 'bucket sync disable' (cc @yuvalif)

so we want to check that `bucket_status.state` still equals `BucketSyncState::Incremental`, and that `bucket_status.incremental_gen` still matches the gen we just finished syncing. if something changed either of those values in the meantime, we should bail out right away. but if we return an error here, this stale generation and shard id would go into the error repo and get retried later - so i think we should avoid that by returning success